### PR TITLE
Return multiple object-safety violation errors and code improvements to the object-safety check

### PR DESCRIPTION
**Pull request by WaffleLapkin**
See individual commits for more information. Split off of #114260, since it turned out that the main intent of that PR was wrong.

r? oli-obk

### CODE REVIEW

File: compiler/rustc_trait_selection/src/traits/object_safety.rs [hunk 1]
```diff
@@ -105,10 +109,9 @@ pub fn is_vtable_safe_method(tcx: TyCtxt<'_>, trait_def_id: DefId, method: ty::A
         return false;
     }
 
-    match virtual_call_violation_for_method(tcx, trait_def_id, method) {
-        None | Some(MethodViolationCode::WhereClauseReferencesSelf) => true,
-        Some(_) => false,
-    }
+    virtual_call_violation_for_method(tcx, trait_def_id, method)
+        .iter()
+        .all(|v| matches!(v, MethodViolationCode::WhereClauseReferencesSelf))
```
Comment (compiler-errors) [inline compiler/rustc_trait_selection/src/traits/object_safety.rs:114]: thanks for doing this, it's so much safer lol

File: compiler/rustc_trait_selection/src/traits/object_safety.rs [hunk 2]
```diff
@@ -109,10 +109,9 @@ pub fn is_vtable_safe_method(tcx: TyCtxt<'_>, trait_def_id: DefId, method: ty::A
         return false;
     }
 
-    match virtual_call_violation_for_method(tcx, trait_def_id, method) {
-        None | Some(MethodViolationCode::WhereClauseReferencesSelf) => true,
-        Some(_) => false,
-    }
+    virtual_call_violation_for_method(tcx, trait_def_id, method)
```
Comment (oli-obk) [inline compiler/rustc_trait_selection/src/traits/object_safety.rs:112]: nit: make `violation` plural in these functions

**Review by oli-obk — APPROVED** [review]
r=me with nit and doc comment fixed

**Pull request merged**